flow/aggregates.py

```diff
@@ -0,0 +1,255 @@
+
+        if num == 1 and sort_by == reverse_order == select is None:
+            setattr(aggregator, '_num', num)
+
+        return cls(aggregator, sort_by, reverse_order, select)
+
+    @classmethod
+    def groupby(cls, key, default=None, sort_by=None, reverse_order=False, select=None):
+        if isinstance(key, str):
+            if default is None:
+                def keyfunction(job):
+                    return job.sp[key]
+            else:
+                def keyfunction(job):
+                    return job.sp.get(key, default)
+
+        elif isinstance(key, Iterable):
+            keys = list(key)
+
+            if default is None:
+                def keyfunction(job):
+                    return [job.sp[key] for key in keys]
+            else:
+                if isinstance(default, Iterable):
+                    if len(default) != len(keys):
+                        raise ValueError("Expected length of default argument is "
+                                         f"{len(keys)}, got {len(default)}.")
+                else:
+                    raise TypeError("Invalid default argument. Expected Iterable, "
+                                    f"got {type(default)}")
+
+                def keyfunction(job):
+                    return [job.sp.get(key, default_value)
+                            for key, default_value in zip(keys, default)]
+
+        elif callable(key):
+            keyfunction = key
+
+        else:
+            raise TypeError("Invalid key argument. Expected either str, Iterable "
+                            f"or a callable, got {type(key)}")
+
+        def aggregator(jobs):
+            for key, group in groupby(sorted(jobs, key=keyfunction), key=keyfunction):
+                yield group
+
+        return cls(aggregator, sort_by, reverse_order, select)
+
+    def __eq__(self, other):
+        return self._aggregator.__code__.co_code == other._aggregator.__code__.co_code and \
+               self._select.__code__.co_code == other._select.__code__.co_code and \
+               self._sort_by == other._sort_by and self._reverse_order == other._reverse_order
+
+    def _create_StoreAggregates(self):
+        # We create the instance of _StoreAggregates or _StoreDefaultAggregates classes
+        # to make sure that we don't explicitly store the aggregates in the FlowProject.
+        # When iterated over these classes, an aggregate is yielded.
+        if not self._is_aggregate:
+            return _StoreDefaultAggregate()
+        else:
+            return _StoreAggregates(self)
+
+    def __call__(self, func=None):
+        if callable(func):
+            setattr(func, '_flow_aggregate', self._create_StoreAggregates())
+            return func
+        else:
+            raise TypeError('Invalid argument passed while calling '
+                            'the aggregate instance. Expected a callable, '
+                            f'got {type(func)}.')
+
+
+class _StoreAggregates:
```

Comment:
Can we just combine the classes? I'm not seeing much benefit to the separation. Especially if you change the `_StoreAggregates` class to compute the aggregates on construction, then both this and the `*Default` class can just store `self._aggregates` (which would just be the project for the default case) and `yield from` that in `__iter__`. Hashing should also work transparently.

Reply:
We could if this implementation had the complete desired behavior. However, once we add `__getitem__` and `__contains__` both of which would look different for the classes and the fact that for the default case the regeneration of aggregates is a no-op. I think having separate classes is warranted.

Reply:
@b-butler The two classes, though does the same thing, are technically not at all similar.
Even if we create `_AggregateStoreBase` we'd then have to override nearly all the methods. (Because even `_register_aggregates` for both the classes registers aggregates differently.